perftests/send-zeros/main.c

```diff
@@ -0,0 +1,139 @@
+#include <sys/socket.h>
+#include <sys/time.h>
+
+#include <fcntl.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <unistd.h>
+
+#include "monoclock.h"
+#include "parsenum.h"
+#include "sock.h"
+#include "warnp.h"
+
+int
+main(int argc, char ** argv)
+{
+	/* Command-line parameters. */
+	const char * addr;
+	size_t buflen;
+	size_t count;
+
+	/* Working variables. */
+	struct sock_addr ** sas_t;
+	struct timeval begin, end;
+	double duration_s;
+	char * buffer;
+	int socket;
+	size_t to_send;
+
+	WARNP_INIT;
+
+	/* Parse command-line arguments. */
+	if (argc != 4) {
+		warn0("usage: %s ADDRESS BUFLEN COUNT", argv[0]);
+		goto err0;
+	}
+	addr = argv[1];
+	if (PARSENUM(&buflen, argv[2], 1, SSIZE_MAX)) {
+		warnp("parsenum");
+		goto err0;
+	}
+	if (PARSENUM(&count, argv[3])) {
+		warnp("parsenum");
+		goto err0;
+	}
+
+	/* Sanity check */
+	if (count == 0) {
+		warnp("Can't send 0 blocks!");
+		goto err0;
+	}
+
+	/* Allocate and fill buffer to send. */
+	if ((buffer = malloc(buflen)) == NULL) {
+		warnp("Out of memory");
+		goto err0;
+	}
+	memset(buffer, 0, buflen);
+
+	/* Initialize count. */
+	to_send = count;
+
+	/* Resolve target address. */
+	if ((sas_t = sock_resolve(addr)) == NULL) {
+		warnp("Error resolving socket address: %s", addr);
+		goto err1;
+	}
+	if (sas_t[0] == NULL) {
+		warn0("No addresses found for %s", addr);
+		goto err2;
+	}
+
+	/* Connect to target. */
+	if ((socket = sock_connect(sas_t)) == -1) {
+		warnp("sock_connect");
+		goto err2;
+	}
+
+	/* Make it blocking. */
+	if (fcntl(socket, F_SETFL,
+	    fcntl(socket, F_GETFL, 0) & (~O_NONBLOCK)) == -1) {
+		warnp("Cannot make connection blocking");
+		goto err3;
+	}
+
+	/* Get beginning time. */
+	if (monoclock_get(&begin)) {
+		warn0("monoclock_get");
+		goto err3;
+	}
+
+	/* Send data. */
+	while (to_send > 0) {
+		if (write(socket, buffer, buflen) != (ssize_t)buflen) {
+			warnp("write failed");
+			goto err3;
+		}
+		to_send--;
+	}
+
+	/* Get ending time. */
+	if (monoclock_get(&end)) {
```

Comment:
This measures how long it takes to *send* the data, but not how long it takes for the data to be received -- it might be sitting in kernel buffers.

I think we discussed having a `shutdown(..., SHUT_WR)` and then calling `read` which will return with 0 (aka EOF) when the remote end is closed?

Reply:
looks like we implemented that in nc-client, but I didn't remember to adjust send-zeros accordingly.